Compile_tb.py generate options not recogised by xrun

<!--
Thank you for your interest in Ibex. We are continuously improving Ibex, and your bug report will help us with that. To help us understand and reproduce your problem as quickly as possible we have prepared a little template below.

No text in blocks starting with < !-- will be shown in the final bug report.

You can also copy/paste images into the bug report, and you can attach files to it by simply dragging them into the issue window.
-->

## Observed Behavior
<!--
Please describe the problem you saw. Include waveforms (in VCD or Verilator/GtkWave FST format only please) and log files if possible.
-->
[compile_tb_stdstreams.log](https://github.com/user-attachments/files/16634707/compile_tb_stdstreams.log)

## Expected Behavior
<!-- Please describe the behavior you expected to see instead. -->

## Steps to reproduce the issue
<!--
Please help us to reproduce the issue you're seeing by including a detailed description of what you did to get to the problem.

For example:
* What command did you run? (Also include the output of the command.)
* What parameters/defines did you set?
* Did you make modifications to the source code?
-->

## My Environment

<!--
Let us know how you're using Ibex. This helps us to better understand your problem and find a solution more quickly.
-->

**EDA tool and version:**
<!--
If you are using an EDA tool to work with Ibex let us know which tool and which version you're using.
For example, "Xilinx Vivado 2020.1" or "Synopsys VCS 2020.03-SP1"
--> Xcelium 23.09-s001

**Operating system:**
<!--
Please let us know the operating system/Linux distribution you're using.
For example "Ubuntu Linux 18.04" or "CentOS 7.3".
--> Centos 7.9

**Version of the Ibex source code:**
<!--
Which version of the Ibex source code are you using?
You can call `git rev-parse HEAD` in your Ibex source directory to get this information.
Please also let us know if you have made changes to the code.
--> 668233699df9ec2a40413e69e0de0a5b10185980

Hi @hcallahan-lowrisc, I rebuilt it in a different directory:
```sh
echo $PKG_CONFIG_PATH
/opt/lib/pkgconfig
```
And still get the same result:
```sh
xrun: *SE,BDOPT: Unknown option -rpath,/opt/lib.
Running compile_tb command :
xrun -64bit -q -f /home/usr14/ibex/dv/uvm/core_ibex/ibex_dv.f -sv -licqueue -uvm -uvmhome CDNS-1.2 -define UVM_REGEX_NO_DPI -elaborate -l /home/usr14/ibex/dv/uvm/core_ibex/out/build/tb/compile_tb.log -xmlibdirpath /home/usr14/ibex/dv/uvm/core_ibex/out/build/tb -access rw -defparam core_ibex_tb_top.RV32E=0 -define IBEX_CFG_RV32M=ibex_pkg::RV32MSingleCycle -define IBEX_CFG_RV32B=ibex_pkg::RV32BOTEarlGrey -define IBEX_CFG_RegFile=ibex_pkg::RegFileFF -defparam core_ibex_tb_top.BranchTargetALU=1 -defparam core_ibex_tb_top.WritebackStage=1 -defparam core_ibex_tb_top.ICache=1 -defparam core_ibex_tb_top.ICacheECC=1 -defparam core_ibex_tb_top.ICacheScramble=1 -defparam core_ibex_tb_top.BranchPredictor=0 -defparam core_ibex_tb_top.DbgTriggerEn=1 -defparam core_ibex_tb_top.SecureIbex=1 -defparam core_ibex_tb_top.PMPEnable=1 -defparam core_ibex_tb_top.PMPGranularity=0 -defparam core_ibex_tb_top.PMPNumRegions=16 -defparam core_ibex_tb_top.MHPMCounterNum=10 -defparam core_ibex_tb_top.MHPMCounterWidth=32 -f ibex_dv_cosim_dpi.f -I/home/usr14/ibex/dv/cosim -lriscv -lsoftfloat -ldl -ldisasm -lfdt -lfesvr -L/opt/lib -I/opt/include -I/opt/include/softfloat -Wld,-Xlinker -rpath,/opt/lib -lstdc++
```
Im pretty sure i didn't specify the prefix at the first time i built it (which is not a smart move). 
For more information, i also check the version of riscv-isa-sim
```sh
 ❯ git rev-parse HEAD
39612f93837122a980395487f55b2d97d29d70c1
❯ git rev-parse --abbrev-ref HEAD
ibex_cosim
```  
Everything seem fine.
I also tried to edit compile.py to use -Wld option of xcelium and it seem to work, howerver the default g++ of xcelium can't understand -rpath option:
```sh
ake[2]: Entering directory `/home/usr14/ibex/dv/uvm/core_ibex'
building library run.so
g++: error: unrecognized command line option '-rpath'
make[2]: *** [/home/usr14/ibex/dv/uvm/core_ibex/out/build/tb/xcelium.d/run.lnx8664.23.09.d/librun.so] Error 1
make[2]: Leaving directory `/home/usr14/ibex/dv/uvm/core_ibex'
```
Can you tell me what to do next? Best regard.

--- a/dv/uvm/core_ibex/scripts/compile_tb.py
+++ b/dv/uvm/core_ibex/scripts/compile_tb.py
@@ -85,6 +85,31 @@ def _main() -> int:
         md.dir_tb.mkdir(exist_ok=True, parents=True)
         md.tb_build_log = md.dir_tb/'compile_tb.log'
 
+        # Locate the spike .pc files to allow us to link against it when
+        # building, riscv-fesvr isn't strictly required but the DV flow has been
+        # observed to see build failures where it isn't present with CentOS 7.
+        spike_iss_pc = ['riscv-riscv', 'riscv-disasm', 'riscv-fdt',
+            'riscv-fesvr']
+        try:
+            subprocess.check_output(['pkg-config', '--exists'] + spike_iss_pc)
+        except subprocess.CalledProcessError as err:
+            raise RuntimeError(
+                f'Failed to find {spike_iss_pc} pkg-config packages. '
+                f'Did you set the PKG_CONFIG_PATH correctly?') from err
+
+        # Now call out to pkg-config to return the appropriate flags for compilation
+        # (The keys here are the substitution placeholders in rtl_simulation.yaml)
+        iss_pkgconfig_dict = {
+            'ISS_CFLAGS'  : ['--cflags'],
+            'ISS_LDFLAGS' : ['--libs-only-other'],
+            'ISS_LIBS'    : ['--libs-only-l', '--libs-only-L'],
+        }
+        iss_cc_subst_vars_dict = \
+            {k: _get_iss_pkgconfig_flags(v, spike_iss_pc, md.simulator)
+             for k, v in iss_pkgconfig_dict.items()}
+
+        # Populate the entire set of variables to substitute in the templated
+        # compilation command, including the compiler flags for the ISS.
         subst_vars_dict = {
             'core_ibex': md.ibex_dv_root,
             'tb_dir': md.dir_tb,
@@ -95,30 +120,7 @@ def _main() -> int:
             'xlm_cov_cfg_file': f"{md.ot_xcelium_cov_scripts}/cover.ccf",
             'dut_cov_rtl_path': md.dut_cov_rtl_path
         }
-
-        # Locate the spike .pc files to allow us to link against it when
-        # building, riscv-fesvr isn't strictly required but the DV flow has been
-        # observed to see build failures where it isn't present with CentOS 7.
-        spike_iss_pc = ['riscv-riscv', 'riscv-disasm', 'riscv-fdt',
-            'riscv-fesvr']
-
-        iss_pkgconfig_dict = {
-            'ISS_CFLAGS'  : ['--cflags'],
-            'ISS_LDFLAGS' : ['--libs-only-other'],
-            'ISS_LIBS'    : ['--libs-only-l', '--libs-only-L'],
-        }
-        md.envvar_PKG_CONFIG_PATH = dict(os.environ).get('PKG_CONFIG_PATH')
-        try:
-            subprocess.check_output(['pkg-config', '--exists'] + spike_iss_pc)
-        except subprocess.CalledProcessError as err:
-            raise RuntimeError(
-                f'Failed to find {spike_iss_pc} pkg-config packages. '
-                f'Did you set the PKG_CONFIG_PATH correctly?') from err
-        subst_vars_dict.update(
-            {k: _get_iss_pkgconfig_flags(v,
-                                         spike_iss_pc,
-                                         md.simulator)
-             for k, v in iss_pkgconfig_dict.items()})
+        subst_vars_dict.update(iss_cc_subst_vars_dict)
 
         md.tb_build_stdout = md.dir_tb/'compile_tb_stdstreams.log'
         md.tb_build_cmds = riscvdv_interface.get_tool_cmds(

--- a/dv/uvm/core_ibex/scripts/compile_tb.py
+++ b/dv/uvm/core_ibex/scripts/compile_tb.py
@@ -5,7 +5,9 @@
 # SPDX-License-Identifier: Apache-2.0
 
 import argparse
+from typing import List
 import os
+import shlex
 import sys
 import subprocess
 import pathlib3x as pathlib
@@ -19,23 +21,51 @@ import logging
 logger = logging.getLogger(__name__)
 
 
-def _get_iss_pkgconfig_flags(specifiers, iss_pc, simulator):
+def _get_iss_pkgconfig_flags(specifiers: List[str], iss_pc: List[str], simulator: str) -> str:
+    all_tokens = []
+
     # Seperate pkg-config calls for each specifier as combining them has been
     # observed misbehaving on CentOS 7
-    _flags = ' '.join([subprocess.check_output(
-                        args=(['pkg-config', s] + iss_pc),
-                        universal_newlines=True,).strip()
-                       for s in specifiers])
+    # Generate a list of tokens for each call, and append it to the all_tokens variable
+    for s in specifiers:
+        cmd = ['pkg-config', s] + iss_pc
+        stdout = subprocess.check_output(cmd, universal_newlines=True).strip()
+        tokens = shlex.split(stdout)
+        logger.debug(f"pkgconfig_tokens = {tokens}")
 
-    if simulator == 'xlm':
-        # See xcelium documentation for the -Wld syntax for passing
-        # flags to the linker. Passing -rpath,<path> options is tricky
-        # because commas are parsed strangely between xrun and the xmsc
-        # tool, and its easy for the options to arrive malformed. Use
-        # the following hack to get it through.
-        if '-Wl' in _flags:  # This should be in LDFLAGS only
-            _flags = "'-Xlinker {}'".format(_flags.replace('-Wl,', ''))
-    return _flags
+        rpath_prefix = '-Wl,-rpath,'
+        def fixup_xcelium_rpath_token(t: str) -> str:
+            """Re-format rpath flags to ensure reliable passing to Xcelium.
+
+            When passing rpath flags to xcelium through the xrun tool, we need to re-format the string
+            output from pkg-config to ensure reliability.
+            This routine detects rpath flags and reformats them according to the Cadence support site
+            article.
+            """
+            if simulator == 'xlm' and t.startswith(rpath_prefix):
+                logger.debug(f"rpath token => {t}")
+                # https://support.cadence.com/apex/ArticleAttachmentPortal?id=a1Od0000000sdF8EAI
+                # See xcelium documentation for the -Wld syntax for passing
+                # user specified arguments to the C++ linker.
+                # Passing -rpath,<path> options is tricky, so use the following workaround as
+                # suggested in the support article.
+                # INPUT:  '-Wl,-rpath,/opt/spike/lib'
+                # OUTPUT: '-Wld,-Xlinker,-rpath,-Xlinker,/opt/spike/lib',
+                rpaths = (t[len(rpath_prefix):]).split(',')
+                xlinker_rpaths_str = ','.join((f"-Xlinker,{p}" for p in rpaths))
+                rpath_token = f"-Wld,-Xlinker,-rpath,{xlinker_rpaths_str}"
+                logger.debug(f"new rpath token => {rpath_token}")
+                return rpath_token
+            else:
+                return t
+
+        tokens = map(fixup_xcelium_rpath_token, tokens)
+
+        all_tokens += tokens
+
+    flags_str = shlex.join(all_tokens)
+    logger.debug(f"flags = {flags_str}")
+    return flags_str
 
 
 def _main() -> int:

--- a/dv/uvm/core_ibex/yaml/rtl_simulation.yaml
+++ b/dv/uvm/core_ibex/yaml/rtl_simulation.yaml
@@ -256,7 +256,7 @@
       <ISS_LIBS>
       <ISS_CFLAGS>
       <EXTRA_COSIM_CFLAGS>
-      -Wld,<ISS_LDFLAGS>
+      <ISS_LDFLAGS>
       -lstdc++
   sim:
     cmd:

